content/integrations/integrating-npm-with-external-services/docker-and-private-modules.mdx

```diff
@@ -4,87 +4,71 @@ redirect_from:
 
 ## Background: runtime variables
 
-You cannot install private npm packages in a Docker container using only runtime variables.  Consider the following Dockerfile:
+You cannot install private npm packages in a Docker container using only runtime variables. Consider the following Dockerfile:
 
 ```
 FROM node
 
-COPY package.json package.json  
+COPY package.json package.json
 RUN npm install
 
 # Add your source files
-COPY . .  
-CMD npm start  
+COPY . .
+CMD npm start
 ```
 
 Which will use the official [Node.js](https://hub.docker.com/_/node) image, copy the `package.json` into our container, installs dependencies, copies the source files and runs the start command as specified in the `package.json`.
 
 In order to install private packages, you may think that we could just add a line before we run `npm install`, using the [ENV parameter](https://docs.docker.com/engine/reference/builder/#env):
 
-```
+```docker
 ENV NPM_TOKEN=00000000-0000-0000-0000-000000000000
 ```
 
 However, this doesn't work as you would expect, because you want the npm install to occur when you run `docker build`, and in this instance, `ENV` variables aren't used, they are set for runtime only.
 
-Instead of run-time variables, you must use a different way of passing environment variables to Docker, available since Docker 1.9: the [ARG parameter](https://docs.docker.com/engine/reference/builder/#arg).
-
-## Create and check in a project-specific .npmrc file
-
-A complete example that will allow you to use `--build-arg` to pass in your NPM_TOKEN requires adding a `.npmrc` file to the project.
+Instead of run-time variables, you must use Docker build secrets.
 
-Use a project-specific `.npmrc` file with a variable for your token to securely authenticate your Docker image with npm.
-
-1. In the root directory of your project, create a custom <a href="https://docs.npmjs.com/cli-documentation/files/npmrc">`.npmrc`</a> file with the following contents:
+## Update the Dockerfile
 
-   ```
-   //registry.npmjs.org/:_authToken=${NPM_TOKEN}
-   ```
+The Dockerfile that takes advantage of this has a few more lines in it than the earlier example that allows us to use your global `.npmrc` and the access token created when running `npm login` command (if you haven't run it already - do so before moving on).
 
-   **Note:** that you are specifying a literal value of `${NPM_TOKEN}`.  The npm cli will replace this value with the contents of the `NPM_TOKEN` environment variable.  Do **not** put a token in this file.
+```dockerfile
+# https://docs.npmjs.com/docker-and-private-modules
+FROM node:16
 
-2. Check in the `.npmrc` file.
+ENV APP_HOME="/app"
 
-## Update the Dockerfile
+WORKDIR ${APP_HOME}
 
-The Dockerfile that takes advantage of this has a few more lines in it than the earlier example that allows us to use the `.npmrc` file and the `ARG` parameter:
+COPY package*.json ${APP_HOME}/
 
-```
-FROM node
+RUN --mount=type=secret,id=npm,target=/root/.npmrc npm install
```

Comment:
in this example you are looking for `/root/.npmrc` correct? On most system the `.npmrc` is going to found in `~/.npmrc`.

Will this above command work for those installations out of the box?

Reply:
> in this example you are looking for `/root/.npmrc` correct?

Yes, that's true.

I've only tested it on macOS and 
```ts
RUN --mount=type=secret,id=npmrc,target=~/.npmrc npm install
``` 
does not work.

AFAIA `/root/.npmrc` will resolve to `~/.npmrc` regardless of your operating system.

```ts
RUN --mount=type=secret,id=npmrc,target=.npmrc npm install
```
will also work and will take your `.npmrc` at the root if the local `.npmrc` is not present. 

Given that we want people to use the `.npmrc` at the root (`~`) where their npm token - generated with `npm login` - is, I think 
```ts
RUN --mount=type=secret,id=npmrc,target=/root/.npmrc npm install
```
is the way to go as it will force people to use the token generated with `npm login` and will not allow for silly behaviour when someone creates a local `.npmrc` and commits their npm token to the repo.